keps/sig-node/2008-checkpoint-restore/README.md

```diff
@@ -0,0 +1,609 @@
+* https://github.com/kubernetes/kubernetes/pull/97689
+
+During discussions with the SIG Node members one possible approach to
+introduce *checkpoint* and *restore* to the CRI API is to provide a new
+*experimental* API. The idea of the *experimental* API is to have an easy
+way to introduce new features without the danger of breaking existing
+functionality. The *experimental* API also makes it easier to rework an API
+as the *experimental* API does not have to be stable.
+
+The existing proof of concept implementation as seen in the listed pull
+requests implements `kubectl drain --checkpoint`. This implementation
+(https://github.com/kubernetes/kubernetes/pull/97194) shows what is necessary
+to implement one of the possible *checkpoint* and *restore* use cases
+(*stateful reboot*). Instead of stopping the containers to drain a node, the
+containers are checkpointed and all state is kept. After draining the node the
+containers will be automatically restored by the kubelet.
+
+The reason to implement the drain use case was that it seemed to be the
+most simple *checkpoint* and *restore* use case. It is not triggered
+by some policy or scheduling decision but only by the user wanting to
+drain a node.
+
+For easier review of the actual changes to the newly introduced
+*experimental* API a second pull request exists with only the
+changes to the API: https://github.com/kubernetes/kubernetes/pull/97689
+
+The proof of concept CRI-O pull request is based on the Podman
+checkpoint/restore implementation which can be seen as a preparation for
+this KEP. Especially as Podman uses CNI for network configuration
+it required changes to CRIU and runc to allow restoring into a
+previously set up network namespace instead of letting CRIU handle
+the network namespace restore:
+
+* https://github.com/opencontainers/runc/pull/1849
+
+This work lead to CRIU's support to allow restoring processes
+out of and into PID namespaces:
+
+* https://github.com/checkpoint-restore/criu/pull/1056
+* https://github.com/opencontainers/runc/pull/2525
+
+As well as the runc support to allow restoring into all kinds of
+namespaces: https://github.com/opencontainers/runc/pull/2583
+
+The support to allow restoring into all kinds of namespaces is
+especially important when checkpointing and restoring containers
+out of and into pods which is possible with the proposed CRI-O changes.
+
+### Implementation
+
+One of the first goals is to implement User Story 1. For this forensic
+use case we want to offer the possibility to checkpoint a container
+out of a running Pod without stopping the checkpointed container and without
+letting the container know that it was checkpointed.
+
+The goal is to introduce *checkpoint* and *restore* in a bottom-up approach.
+In a first step we only want to extend the CRI API to trigger a checkpoint
+by the container engine and to have the low level primitives in the *kubelet*
+to trigger a checkpoint.
```

Comment:
Can we also mention the kubelet API to trigger the checkpoint?

Reply:
Done-